File: bindings/matlab/tests/Test_TriangularMap.m
clear

addpath(genpath('../../../'));

KokkosInitialize(8);

opts = MapOptions(); %Setting options to default
multis1 = [0 1 2 3]';
multis2 = [0 1;2 0]; %need to specify it as (order,dim) !!!
mset1 = MultiIndexSet(multis1);
mset2 = MultiIndexSet(multis2);

map1 = ConditionalMap(mset1,opts); %create ConditionalMap with Matlab MultiIndexSet and MapOptions
map2 = ConditionalMap(mset2,opts);

triMap = TriangularMap([map1,map2]);

disp(map1.Coeffs)
coeffs = 0.1*[1 2 3 4 5 6];
triMap.SetCoeffs(coeffs)
disp(triMap.Coeffs)
disp(map1.Coeffs)

disp('GetComponent:')
map1FromTri = triMap.GetComponent(1);
map2FromTri = triMap.GetComponent(2);

disp(map1FromTri.Coeffs)
disp(map2FromTri.Coeffs)

% Setting new coeff values
map1FromTri.SetCoeffs([1 1 1 1])

disp('GetBaseFunction:')
parFunc = map1FromTri.GetBaseFunction();
parFunc.SetCoeffs(map1FromTri.Coeffs);
disp(parFunc.Evaluate(randn(1,10)))


% Test construction with different dimensions

opts=MapOptions;
maxDegree = 2;

numBlocks = 3;
dim1 = 2;
dim2 = 4;
dim3 = 3;
dim = dim1 + dim2 + dim3;

M1 = CreateTriangular(dim1,dim1,maxDegree,opts);
M2 = CreateTriangular(dim1+dim2,dim2,maxDegree,opts);
M3 = CreateTriangular(dim1+dim2+dim3,dim3,maxDegree,opts);

triMap = TriangularMap([M1 M2 M3]);

triMap.SetCoeffs(randn(dim,triMap.numCoeffs));

disp(['Num coeffs triMap: ',num2str(triMap.numCoeffs)])
disp(['Sum coeffs maps: ',num2str(M1.numCoeffs+M2.numCoeffs+M3.numCoeffs)])

Y=triMap.Evaluate(randn(dim,10));

File: bindings/python/tests/test_TriangularMap.py
# Test methods of TriangularMap object
import mpart
import numpy as np

opts = mpart.MapOptions()

multis_1 = np.array([[0],[1]])  # linear
multis_2 = np.array([[0,0],[0,1],[2,0]])  # quadratic in x_1, linear in x_2, matches form of target 

mset_1 = mpart.MultiIndexSet(multis_1).fix(True)
mset_2 = mpart.MultiIndexSet(multis_2).fix(True)

map_1 = mpart.CreateComponent(mset_1, opts)
map_2 = mpart.CreateComponent(mset_2, opts)

triangular = mpart.TriangularMap([map_1,map_2])
num_samples = 100
x = np.random.randn(2,num_samples)


def test_numCoeffs():
    assert triangular.numCoeffs == 2 + 3


def test_CoeffsMap():
    coeffs = np.random.randn(triangular.numCoeffs)
    triangular.SetCoeffs(coeffs)
    assert np.all(triangular.CoeffMap() == coeffs)


def test_Evaluate():
    assert triangular.Evaluate(x).shape == (2,num_samples)


def test_LogDeterminant():
    assert triangular.LogDeterminant(x).shape == (num_samples,)
    
def test_Inverse():
    coeffs = np.random.randn(triangular.numCoeffs)
    triangular.SetCoeffs(coeffs)
    y = triangular.Evaluate(x)
    x_ = triangular.Inverse(np.zeros((0,num_samples)),y)
    assert np.allclose(x_, x, atol=1E-3)
